controllers/machine_webhook.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"net/http"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// +kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-machine,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machines,verbs=create;update,versions=v1alpha3,name=providervalidation.machine.infrastructure.cluster.x-k8s.io
+
+// MachineValidator validates Machines
+type MachineValidator struct {
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+// Handle validates Machine admission requests.
+func (v *MachineValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	m := &clusterv1.Machine{}
+
+	err := v.decoder.Decode(req, m)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if m.Spec.InfrastructureRef.Kind == "AWSMachine" && m.Spec.Version == nil {
+		err := apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("Machine").GroupKind(), m.Name, field.ErrorList{
+			field.Required(field.NewPath("spec", "version"), "is required"),
+		})
+		return admission.Denied(err.Error())
+	}
```

Comment:
This should probably be added to the existing Machine Webhook here: https://github.com/kubernetes-sigs/cluster-api/blob/83c7edf23e727d89880f3adc697a9db41488e98c/api/v1alpha3/machine_webhook.go#L65-L102

Reply:
Version is optional for CAPI but currently required for CAPA to do the ami lookup (if you omit the image id). I’m not sure it’s appropriate right now to require it in the CAPI webhook but it’s ok to put it in for CAPA. Wdyt?

Reply:
+1, it should be kept in infrastructure providers

Reply:
Note: I think I need to update the webhook to only require version if `spec.ami.id` is not set.